# Урок 2. Анализ датасета с помощью Pandas

## Задача 1: 

**Задание 1**
*Скачать данные по ссылке https://www.kaggle.com/datasets/esratmaria/house-price-dataset-with-other-information
Считать данные с помощью pandas
Вывести на экран первые 5 строк
Посмотреть на описание признаков и на их содержание*

**Решение**

*По заданной ссылке скачан файл kc_house_data.csv*

*Подключаем pandas и задаём директорию (выполнение в ОС Ubuntu), содержащую файл с данными:*

In [3]:
import pandas as pd
dir = "~/Рабочий стол/GeekBrains/Python for analysts/PforAnalys"

*Считываем данные из файла с помощью pandas:*

In [4]:
df = pd.read_csv(dir + "/kc_house_data.csv")

*Выводим первые 5 строк файла:*

In [10]:
df.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...      7        1180              0   
1      7242     2.0           0     0  ...      7        2170            400   
2     10000     1.0           0     0  ...      6         770              0   
3      5000     1.0           0     0  ...      7        1050            910   
4      8080     1.0           0     0  ...      8        1680              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1955             0    98178  47.5112 -122.257           1340   
1      1951          1991    98125  47.7210 -122.319           1690   
2      1933             0    98028  47.7379 -122.233           2720   
3      1965             0    98136  47.5208 -122.393           1360   
4      1987             0    98074  47.6168 -122.045           1800   

   sqft_lot15  
0        5650  
1        7639  
2        8062  
3        5000  
4        7503  

[5 rows x 21 columns]

*По содержимому видно, что каждая строка содержит данные по квартире: стоимость, квадратные метры и т.д. Всего 21 параметр*

## Задача 2

**Задача 1**

*Проведите первичный анализ данных  
Изучите типы данных  
Найдите количество пропущенных ячеек в данных  
Посчитайте основные статистики по всем признакам и поизучайте их  
Пишите выводы*

***Смотрим общую информацию по датафрейму:***

In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             21613 non-null  int64  
 1   date           21613 non-null  object 
 2   price          21613 non-null  float64
 3   bedrooms       21613 non-null  int64  
 4   bathrooms      21613 non-null  float64
 5   sqft_living    21613 non-null  int64  
 6   sqft_lot       21613 non-null  int64  
 7   floors         21613 non-null  float64
 8   waterfront     21613 non-null  int64  
 9   view           21613 non-null  int64  
 10  condition      21613 non-null  int64  
 11  grade          21613 non-null  int64  
 12  sqft_above     21613 non-null  int64  
 13  sqft_basement  21613 non-null  int64  
 14  yr_built       21613 non-null  int64  
 15  yr_renovated   21613 non-null  int64  
 16  zipcode        21613 non-null  int64  
 17  lat            21613 non-null  float64
 18  long  

*Используемые типы данных: 5 полей вещественного типа, 15 полей целочисленного, 1 поле как object - дата*

***Количество пустых ячеек:***

In [12]:
df.isnull().sum().sum()

0

*Хотя то, что пропущенных ячеек нет, видно из предыдущего шага*

***Основные статистические характеристики:***

In [13]:
df.describe()

id         price      bedrooms     bathrooms   sqft_living  \
count  2.161300e+04  2.161300e+04  21613.000000  21613.000000  21613.000000   
mean   4.580302e+09  5.400881e+05      3.370842      2.114757   2079.899736   
std    2.876566e+09  3.671272e+05      0.930062      0.770163    918.440897   
min    1.000102e+06  7.500000e+04      0.000000      0.000000    290.000000   
25%    2.123049e+09  3.219500e+05      3.000000      1.750000   1427.000000   
50%    3.904930e+09  4.500000e+05      3.000000      2.250000   1910.000000   
75%    7.308900e+09  6.450000e+05      4.000000      2.500000   2550.000000   
max    9.900000e+09  7.700000e+06     33.000000      8.000000  13540.000000   

           sqft_lot        floors    waterfront          view     condition  \
count  2.161300e+04  21613.000000  21613.000000  21613.000000  21613.000000   
mean   1.510697e+04      1.494309      0.007542      0.234303      3.409430   
std    4.142051e+04      0.539989      0.086517      0.766318      0.650743   
min    5.200000e+02      1.000000      0.000000      0.000000      1.000000   
25%    5.040000e+03      1.000000      0.000000      0.000000      3.000000   
50%    7.618000e+03      1.500000      0.000000      0.000000      3.000000   
75%    1.068800e+04      2.000000      0.000000      0.000000      4.000000   
max    1.651359e+06      3.500000      1.000000      4.000000      5.000000   

              grade    sqft_above  sqft_basement      yr_built  yr_renovated  \
count  21613.000000  21613.000000   21613.000000  21613.000000  21613.000000   
mean       7.656873   1788.390691     291.509045   1971.005136     84.402258   
std        1.175459    828.090978     442.575043     29.373411    401.679240   
min        1.000000    290.000000       0.000000   1900.000000      0.000000   
25%        7.000000   1190.000000       0.000000   1951.000000      0.000000   
50%        7.000000   1560.000000       0.000000   1975.000000      0.000000   
75%        8.000000   2210.000000     560.000000   1997.000000      0.000000   
max       13.000000   9410.000000    4820.000000   2015.000000   2015.000000   

            zipcode           lat          long  sqft_living15     sqft_lot15  
count  21613.000000  21613.000000  21613.000000   21613.000000   21613.000000  
mean   98077.939805     47.560053   -122.213896    1986.552492   12768.455652  
std       53.505026      0.138564      0.140828     685.391304   27304.179631  
min    98001.000000     47.155900   -122.519000     399.000000     651.000000  
25%    98033.000000     47.471000   -122.328000    1490.000000    5100.000000  
50%    98065.000000     47.571800   -122.230000    1840.000000    7620.000000  
75%    98118.000000     47.678000   -122.125000    2360.000000   10083.000000  
max    98199.000000     47.777600   -121.315000    6210.000000  871200.000000

***Выводы:***

*С помощью библиотеки pandas в одну строку можно считать csv-файл. При этом происходит автоопределение типов данных, правда это касается числовых типов - все остальные определяются как object, которые можно самостоятельно переопределить при необходимости. Функция info() показывает структуру датафрейма: количество столбцов, строк, типыданных ячеек, количество пустых ячеек, общий объём загруженных данных*

## Задача 3

**Ответьте на несколько вопросов**

*В каком диапазоне изменяются стоимости недвижимости?*

**Находим минимальное значение цены:**

In [14]:
df['price'].min()

75000.0

**Находим максимальное значение цены:**

In [15]:
df['price'].max()

7700000.0

### -------------------------------------------------------------

*Какую долю в среднем занимают жилая площадь от всей площади по всем домам?*

**Ответ:**

*Здесь sqft_living - жилая площадь, sqft_lot - общая площадь. Для ответа на вопрос необходимо найти отношение среднего значения жилой площади к среднему значению общей площади:*

In [16]:
df['sqft_living'].mean() / df['sqft_lot'].mean()

0.1376781757959227

### -------------------------------------------------------------

*Как много домов с разными этажами в данных?*

**Ответ:**

*Здесь используем поле floors (этажи)*

In [17]:
df['floors'].value_counts()

1.0    10680
2.0     8241
1.5     1910
3.0      613
2.5      161
3.5        8
Name: floors, dtype: int64

*И видим статистику по этажности и количеству: одноэтажных домов - 10680, двухэтажных - 8241 и т.д.*  
*Раньше не встречался с полутораэтажными строениями, поэтому загуглил.* ***1.5-этажный дом – это строение, в котором высота наружных стен выше уровня первого этажа, но не дотягивает до полноценной высоты наружных стен второго этажа.*** *По аналогии - с этажностью домов 2,5 и 3,5*

### -------------------------------------------------------------

*Насколько хорошие состояния у домов в данных?*

**Ответ:**

*Работаем с полем condition:*

*Мнимальное значение состояние, встречаемое в данных - 1, максимальное - 5. Т.е. состояние оценивается по пятибальной системе. Находим среднее значение поля condition:*

In [18]:
df['condition'].mean()

3.4094295100171195

*Но не стоит усреднять состояние. Лучше выполнить более детальную выборку:*

In [20]:
df['condition'].value_counts(normalize = True)

3    0.649193
4    0.262759
5    0.078703
2    0.007958
1    0.001388
Name: condition, dtype: float64

*Получаем: 7,8% в отличном состоянии, 26,3% - в хорошем, 65% - в удовлетворительном, 0,8% в плохом состоянии, 0,1% в ужасном*

### -------------------------------------------------------------

*Найдите года, когда построили первый дом, когда построили последний дом в данных?*

**Ответ:**

*Работаем с полем date. В данный момент оно определено как object. Чтобы работать как с датой необходимо преобразовать тип данных данного поля*

In [24]:
df['date'].dtype

dtype('O')

*Да, тип данных не определён. Требуется преобразование*

In [25]:
df['date'] = pd.to_datetime(df['date'])

*И проверяем:*

In [26]:
df['date'].dtype

dtype('<M8[ns]')

*Первый дом был построен:*

In [28]:
df['date'].min()

Timestamp('2014-05-02 00:00:00')

*Т.е. первый дом из выборки данных был построен 2 мая 2014 года*

*Последний дом был построен:*

In [29]:
df['date'].max()

Timestamp('2015-05-27 00:00:00')

*Последний дом из выборки данных был построен 27 мая 2015 года*

### -------------------------------------------------------------

## Задача 4

**Ответьте на несколько вопросов**

*Сколько в среднем стоят дома, у которых 2 спальни?*

**Ответ:**

*Фильтруем выборку по полю bedrooms:*

In [9]:
bedrooms = df[df['bedrooms'] == 2]

*И находим среднее значение поля price*

In [12]:
bedrooms['price'].mean()

401372.681884058

### ------------------------------------------------------------------------------------

*Какая в среднем общая площадь домов, у которых стоимость больше 600 000?*

**Ответ:**

*Фильтруем дома стоимостью выше 600 000*

In [19]:
houses = df[df['price'] > 600000]

*Находим усреднённую общую площадь (кв. футов):*

In [20]:
houses['sqft_lot'].mean()

20442.524776214832

### ------------------------------------------------------------------------------------

*Как много домов коснулся ремонт?*

**Ответ:**

*Смотрим поле yr_renovated - год ремонта. И отбираем дома, у которых год ремонта не равен нулю:*

In [15]:
houses = df[df['yr_renovated'] != 0]

*И определяем их количество:*

In [16]:
houses['id'].count()

914

### ------------------------------------------------------------------------------------

*Насколько в среднем стоимость домов с оценкой grade домов выше 10 отличается от стоимости домов с оценкой grade меньше 4?*

**Ответ:**

*Фильтруем дом с классом (grade) выше 10 и с классом ниже 4 в два датафрейма:*

In [17]:
grade10 = df[df['grade'] > 10]
grade4 = df[df['grade'] < 4]

*И находим разницу средних значений цены для двух категорий классов:*

In [18]:
grade10['price'].mean() - grade4['price'].mean()

1488885.1175298805

## Задача 5

*5.1 Выберите дом клиенту  
Клиент хочет дом с видом на набережную, как минимум с тремя ванными и с подвалом. Сколько вариантов есть у клиента?*

**Решение**

*Формируем выборку по трём условиям. Причём все три условия обязательны к выполнению, т.е. используем логическое И: 
waterfront (береговая линия) не равно нулю (т.е. присутствует), bathrooms равно трём, sqft_basement (площадь подвала) не равна нулю*

In [21]:
houses = df[(df['waterfront'] == 1) & (df['bathrooms'] == 3) & (df['sqft_basement'] != 0)]

In [22]:
houses['id'].count()

5

*У клиента 5 вариантов по указанным условиям*

### ------------------------------------------------------------------------------------

*5.2 Выберите дом клиенту  
Клиент хочет дом либо с очень красивым видом из окна, либо с видом на набережную, в очень хорошем состоянии и год постройки не меньше 1980 года. В каком ценовом диапазоне будут дома?*

**Решение**

*Формируем выборку по условиям: view равно 4 (4 - максимальное значение у вида, т.е. самый красивый) ИЛИ waterfront не равно нулю И condition равно 5 и yr_built больше или равно 1980:*

In [23]:
houses = df[((df['waterfront'] == 1) | (df['view'] == 4)) & (df['condition'] == 5) & (df['yr_built'] >= 1980)]

*Определяем минимальную и максимальную цены в полученной выборке:*

In [24]:
df['price'].min()

75000.0

In [25]:
df['price'].max()

7700000.0

*По заданным условиям ценовой диапазон составляет от 75 000 до 7 700 000*

### ------------------------------------------------------------------------------------

*5.3 Выберите дом клиенту  
Клиент хочет дом без подвала, с двумя этажами, стоимостью до 150000. Какая оценка по состоянию у таких домов в среднем?*

**Решение**

*Формируем выборку по условиям: sqft_basement = 0 (площадь подвала нулевая), floors (этажи) = 2, price (цена) меньше или равно 150 000:*

In [28]:
houses = df[(df['sqft_basement'] == 0) & (df['floors'] == 2) & (df['price'] <= 150000)]

*Находим среднюю оценку состояния:*

In [29]:
houses['condition'].mean()

3.0

*В среднем оценка состояния - 3, удовлетворительное состояние*